sw/device/tests/csrng_edn_concurrency_test.c

```diff
@@ -0,0 +1,392 @@
+
+  entropy_testutils_stop_all();
+  CHECK_DIF_OK(dif_entropy_src_configure(
+      &entropy_src, entropy_testutils_config_default(), kDifToggleEnabled));
+  CHECK_DIF_OK(dif_csrng_configure(&csrng));
+  CHECK_DIF_OK(dif_edn_set_auto_mode(&edn0, edn_params0));
+  CHECK_DIF_OK(dif_edn_set_auto_mode(&edn1, edn_params1));
+}
+
+/**
+ * Updates and logs the execution state.
+ *
+ * @param next_state The next state.
+ */
+static void execution_state_update(test_state_t next_state) {
+  CHECK(next_state < kTestStateCount);
+  LOG_INFO("Test state: %s", kStatusNames[next_state]);
+  execution_state = next_state;
+}
+
+/**
+ * Main test task.
+ *
+ * Configures entropy complex and signals test tasks to run an iteration loop.
+ *
+ * @param task_parameters Unused. Set to NULL by ottf.
+ */
+static void main_task(void *task_parameters) {
+  task_iter_count_max[kTestTaskIdMain] = kTestParamNumIterationsSim;
+  if (kDeviceType != kDeviceSimDV && kDeviceType != kDeviceSimVerilator) {
+    task_iter_count_max[kTestTaskIdMain] = kTestParamNumIterationsOther;
+  }
+
+  uint32_t iter_count = 0;
+  while (true) {
+    if (execution_state == kTestStateRun) {
+      uint32_t count = 0;
+      for (size_t i = 0; i < kTestTaskIdCount; ++i) {
+        count += task_done[i] ? 1 : 0;
+      }
+      if (count < kTestTaskIdCount) {
+        ottf_task_yield();
+        continue;
+      }
+      // Consider this test iteration done if all tasks are done. Update the
+      // iteration counter and set the execution state to `kTestStateSetup`
+      // to continue with configuring the entropy blocks for the next test
+      // run iteration.
+      if (++iter_count >= task_iter_count_max[kTestTaskIdMain]) {
+        execution_state_update(kTestStateTearDown);
+        break;
+      }
+      execution_state_update(kTestStateSetup);
+      csrng_testutils_recoverable_alerts_check(&csrng);
+    }
+    // The rest of this code block is executed when
+    // `execute_state == kTestStateSetup`.
+    entropy_config();
+    for (size_t i = 0; i < kTestTaskIdCount; ++i) {
+      task_done[i] = false;
+    }
+    // Signal the other tasks to start test execution.
+    execution_state_update(kTestStateRun);
+    // This is required to ensure the aggregation of task_done entries is equal
+    // to `kTestTaskIdCount`.
+    task_done_set_and_yield(kTestTaskIdMain);
+  }
+  task_terminated[kTestTaskIdMain] = true;
+  OTTF_TASK_DELETE_SELF_OR_DIE;
+}
+
+bool test_main(void) {
+  init_peripherals();
+
+  // Initialize the test `execution_state` before launching the tasks.
+  execution_state_update(kTestStateSetup);
+  for (size_t i = 0; i < kTestTaskIdCount; ++i) {
+    task_terminated[i] = false;
+  }
+
+  CHECK(ottf_task_create(main_task, "main", kOttfFreeRtosMinStackSize, 1));
+  CHECK(ottf_task_create(otbn_task, "otbn", kOttfFreeRtosMinStackSize, 1));
+  CHECK(ottf_task_create(ibex_task, "ibex", kOttfFreeRtosMinStackSize, 1));
```

Comment:
Technically there is nothing wrong with this setup, just dropping an FYI here. The `ottf_task_create` function [automatically increments](https://cs.opensource.google/opentitan/opentitan/+/master:sw/device/lib/testing/test_framework/ottf_main.c;drc=9ae76126b135cbc8889b05a6c74bc9fc2cd5401d;l=67) the task priority you pass by 1, so as to avoid running sub-tasks at the same priority level as this `test_main` task, which has a [priority level of 1](https://cs.opensource.google/opentitan/opentitan/+/master:sw/device/lib/testing/test_framework/ottf_main.c;drc=9ae76126b135cbc8889b05a6c74bc9fc2cd5401d;l=142), and is the first task created (other than FreeRTOS' idle task) when you set the `enable_concurrency` test configuration flag to `true`. Because of that, the sub tasks you spawn should have a priority of 2, meaning the polling for task completion below shouldn't be necessary, as FreeRTOS won't start executing this `test_main` function again (after you call `ottf_task_yield` for the first time) until all subtasks have been deleted. Essentially by nature of the priority levels, you will get an equivalent of a `join` function for free.

Plmk if you think the auto-increment of the priority level by the `ottf_task_create` function is confusion, and I can remove this.

Reply:
A nice. This makes sense. I am going to update the code to remove the check at the end. I guess it would be nice to have something like `ottf_join()` to make this more explicit, even though join would start with a yield.  I will add a comment to clarify the behavior in case someone else uses it as a template for another test case.